quic/s2n-quic-core/src/recovery/bbr.rs

```diff
@@ -95,21 +95,28 @@ enum State {
 
 impl State {
     /// The dynamic gain factor used to scale BBR.bw to produce BBR.pacing_rate
-    fn pacing_gain(&self) -> Ratio<u64> {
+    fn pacing_gain(&self, app_settings: &ApplicationSettings) -> Ratio<u64> {
         match self {
             State::Startup => startup::PACING_GAIN,
             State::Drain => drain::PACING_GAIN,
-            State::ProbeBw(probe_bw_state) => probe_bw_state.cycle_phase().pacing_gain(),
+            State::ProbeBw(probe_bw_state) => {
+                probe_bw_state.cycle_phase().pacing_gain(app_settings)
+            }
             State::ProbeRtt(_) => probe_rtt::PACING_GAIN,
         }
     }
 
     /// The dynamic gain factor used to scale the estimated BDP to produce a congestion window (cwnd)
-    fn cwnd_gain(&self) -> Ratio<u64> {
+    fn cwnd_gain(&self, app_settings: &ApplicationSettings) -> Ratio<u64> {
+        let cwnd_gain = app_settings
+            .cwnd_gain
+            .map(|cwnd_gain| Ratio::new_raw(cwnd_gain as u64, 100))
+            .unwrap_or(probe_bw::CWND_GAIN);
```

Comment:
could these be functions on an `impl ApplicationSettings`?

Reply:
We could but imo it makes things less readable because app_settings represent an override to the default values. We should prioritize co-locating code/comments with the the default implementation.

For example:
- initial_congestion_window: this logic is quite involved and vital to BBR so prob should stay as impl BbrCongestionController.
- probe_bw_cwnd_gain: in this function the more interesting logic seems to be around matching on the State enum
- probe_bw_up_pacing_gain: this is in the probe_bw.rs file and uses the constant UP_PACING_GAIN defined there. Seems like it would be easier to follow along if the logic and compliance comment stayed there
- loss_thresh: this is pretty isolated and could be moved to ApplicationSettings but it would be the only one and that seems odd.

Reply:
I think at least the mapping to a ratio could happen in a function in `ApplicationSettings` though